docs/man/man3/physics.3
.TH "physics" 3 "Fri Apr 27 2018" "Variational Analysis" \" -*- nroff -*-
.ad l
.nh
.SH NAME
physics \- 
.SH SYNOPSIS
.br
.PP
.SS "Functions/Subroutines"

.in +1c
.ti -1c
.RI "subroutine \fBs_r_to_t_z\fP (P, PS, ZS, S, R, T, Z)"
.br
.ti -1c
.RI "subroutine \fBt_r_to_s_z\fP (P, PS, ZS, T, R, S, Z)"
.br
.ti -1c
.RI "subroutine \fBdiverg\fP (UNITH, VAR, U, V, DIVU, DIVV, DIV1)"
.br
.ti -1c
.RI "subroutine \fBfcorlx\fP (UNITH, NSTU, F, V, FCX1)"
.br
.ti -1c
.RI "subroutine \fBfcorly\fP (UNITH, NSTU, F, U, FCY1)"
.br
.ti -1c
.RI "subroutine \fBfpgd\fP (UNITH, Z, DZDX, FP1)"
.br
.ti -1c
.RI "subroutine \fBheight\fP (NP, KS, KT, P, T, R, Z)"
.br
.ti -1c
.RI "subroutine \fBcalht2\fP (PS, ZS, N, P, TC, R, HT, DEW)"
.br
.ti -1c
.RI "subroutine \fBcalc_state2\fP (P, T, TD, E, Z, R, RH, H, HDRY, S)"
.br
.in -1c
.SH "Function/Subroutine Documentation"
.PP 
.SS "subroutine physics::calc_state2 (real (kind=rk8), intent(in) P, real (kind=rk8), intent(in) T, real (kind=rk8), intent(in) TD, real (kind=rk8), intent(in) E, real (kind=rk8), intent(in) Z, real (kind=rk8), intent(out) R, real (kind=rk8), intent(in) RH, real (kind=rk8), intent(out) H, real (kind=rk8), intent(out) HDRY, real (kind=rk8), intent(out) S)"

.PP
Definition at line 368 of file physics\&.f90\&.
.PP
.nf
368 USE portable
369 USE constants
370 
371 IMPLICIT NONE
372 REAL (KIND=RK8), INTENT(IN)     :: p    ! Pressure (Pa) - check units\&.
373 REAL (KIND=RK8), INTENT(IN)     :: t    ! Temperature (K)\&.
374 REAL (KIND=RK8), INTENT(IN)     :: td   ! Dew point temperature (K)\&.
375 REAL (KIND=RK8), INTENT(IN)     :: e    ! Vapour pressure (Pa - check units\&.
376 REAL (KIND=RK8), INTENT(IN)     :: z    ! Geopotential height (m) - check units\&.
377 REAL (KIND=RK8), INTENT(OUT)    :: r    ! Mixing ratio\&.
378 REAL (KIND=RK8), INTENT(IN)     :: rh   ! Relative humidity (decimal fraction between 0 and 1)\&. 
379 REAL (KIND=RK8), INTENT(OUT)    :: h    !
380 REAL (KIND=RK8), INTENT(OUT)    :: hdry !
381 REAL (KIND=RK8), INTENT(OUT)    :: s    !
382 
383 !
384 ! Local variables\&.
385 !
386 REAL (KIND=RK8)                 :: ep, es, rs, lv1
387 
388 ep      = rd/rw
389 es      = 1\&.003*exp(53\&.67957 - 6743\&.769/t - 4\&.8451*log(t))  ! An empirical formula for the saturation vapour pressure\&.
390 rs      = ep*es/(p-es)
391 r       = rs*rh
392 lv1     = lv0 + (cpv-cl)*(t-t0)
393 h       = (cpd + r*cl)*t + lv1*r + (1\&.0 + r)*g*z
394 hdry    = cpd*t + g*z
395 s       = cpd*log(t) - rd*log(p) + lv0*r/t - r*rw*log(rh)
.fi
.SS "subroutine physics::calht2 (real (kind=rk8), intent(in) PS, real (kind=rk8), intent(in) ZS, integer (kind=ik4), intent(in) N, real (kind=rk8), dimension(n), intent(in) P, real (kind=rk8), dimension(n), intent(in) TC, real (kind=rk8), dimension(n), intent(inout) R, real (kind=rk8), dimension(n), intent(out) HT, real (kind=rk8), dimension(n), intent(out) DEW)"

.PP
Definition at line 284 of file physics\&.f90\&.
.PP
.nf
284 USE portable
285 USE constants
286 
287 IMPLICIT NONE
288 REAL (KIND=RK8), INTENT(IN)                       :: ps     ! Surface pressure (units)\&.
289 REAL (KIND=RK8), INTENT(IN)                       :: zs     ! Surface geopotential (units)\&.
290 INTEGER (KIND=IK4), INTENT(IN)                    :: n      ! Number of vertical levels\&.
291 REAL (KIND=RK8), DIMENSION(N), INTENT(IN)         :: p      ! Pressures (units)\&.
292 REAL (KIND=RK8), DIMENSION(N), INTENT(IN)         :: tc     ! Temperatures (Celsius)\&.
293 REAL (KIND=RK8), DIMENSION(N), INTENT(INOUT)      :: r      ! Relative humidity (%)\&.
294 REAL (KIND=RK8), DIMENSION(N), INTENT(OUT)        :: ht     ! Geopotential height (units)\&.
295 REAL (KIND=RK8), DIMENSION(N), INTENT(OUT)        :: dew    ! Dew-point temperature? (units)\&.
296 
297 !
298 ! Local variables\&.
299 !
300 REAL (KIND=RK8), DIMENSION(N)                     :: rh     ! Relative humidity converted to 0-1 range\&.
301 REAL (KIND=RK8), DIMENSION(N)                     :: t      ! Temperature in Kelvin\&.
302 REAL (KIND=RK8), DIMENSION(N)                     :: es     ! Saturation vapour pressure\&.
303 REAL (KIND=RK8), DIMENSION(N)                     :: rs     ! Saturated mixing ratio\&.
304 REAL (KIND=RK8)                                   :: ep     ! = RD/RW
305 INTEGER (KIND=IK4)                                :: ks     ! Index of first vertical level at or above the surface\&.
306 REAL (KIND=RK8)                                   :: r1, dz
307 
308 !
309 ! Convert the relative humidity to the 0-1 range\&. Any humidities less than 0 or greater than 1 are set to 1\&. Not sure of the
310 ! reasoning behind making RH<0 go to 1 \&.\&.\&. maybe missing data is better substituted with 1 RH than some other value?
311 !
312 rh  = r/100\&.0               
313 WHERE ((rh \&.LT\&. 0\&.0) \&.OR\&. (rh \&.GT\&. 1\&.0)) rh = 1\&.0
314 
315 t   = tc + t0               ! Convert temperatures in Celsius to Kelvin\&.
316 
317 !
318 ! Calculate the saturation vapour pressure using a modified form of the Clausius-Clayperon equation\&. I believe this equation might
319 ! be in Emanuel (1994): "Atmospheric Convection" (maybe)\&.
320 !
321 es  = 1\&.003*exp(53\&.67957 - 6743\&.769/t - 4\&.8451*log(t))
322 ep  = rd/rw                 ! This will be used several times in this subroutine (so save resources by only calculating it once)\&.
323 
324 rs  = ep*es/(p-es)          ! Calculate the saturated mixing ratio\&.
325 
326 !
327 ! Overwrite the input relative humidities (R, in %) with mixing ratio\&.
328 !
329 r   = rs*rh
330 
331 !
332 ! Locate the first level, at or above the surface (which has pressure PS)\&. Default to 1 if we can't find the surface\&.
333 !
334 ks  = max(transfer(maxloc(p, mask = p \&.LE\&. ps), 0),1)
335 
336 !
337 ! Work out the geopotential height of the first level\&. I added a little bit of code to deal with the situation where KS=1\&. The IDL
338 ! code did not seem to deal with this, assuming that there was always at least one pressure level "below the surface", and did a
339 ! simple interpolation\&. If there is no level below the surface, can't do interpolation\&.
340 !
341 r1      = rd*0\&.5*((1+r(max(1,ks-1))/ep)/(1+r(max(1,ks-1))) + (1+r(ks)/ep)/(1+r(ks)))
342 dz      = r1/g*0\&.5*(t(max(1,ks-1))/p(max(1,ks-1)) + t(ks)/p(ks)) * (ps/p(ks))
343 ht(ks)  = zs + dz
344 
345 !
346 ! Work out the geopotential height of all the vertical levels above the surface\&.
347 !
348 CALL height(n, ks, n, p, t, r, ht)
349 
350 !
351 ! Assign the surface geopotential height to all levels below the surface\&.
352 !
353 ht(1:max(1,ks-1))  = zs
354 
.fi
.SS "subroutine physics::diverg (real (kind=rk8), dimension(:), intent(in) UNITH, real (kind=rk8), dimension(:,:), intent(in) VAR, real (kind=rk8), dimension(:,:), intent(in) U, real (kind=rk8), dimension(:,:), intent(in) V, real (kind=rk8), dimension(:,:), intent(in) DIVU, real (kind=rk8), dimension(:,:), intent(in) DIVV, real (kind=rk8), dimension(:), intent(out) DIV1)"

.PP
Definition at line 170 of file physics\&.f90\&.
.PP
.nf
170 USE portable
171 
172 IMPLICIT NONE
173 REAL (KIND=RK8), DIMENSION(:), INTENT(IN)     :: unith      ! One dimensional array with y columns\&.
174 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: var        ! Array with x columns and y rows containing the variable\&.
175 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: u          ! Array with x columns and y rows containing U-component of wind\&.
176 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: v          ! Array with x columns and y rows containing V-component of wind\&.
177 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: divu       ! Array with x columns and y rows containing DIVU\&.
178 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: divv       ! Array with x columns and y rows containing DIVV\&.
179 REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)    :: div1       ! Array with x columns containing divergence\&.
180 
181 div1    = matmul((u*var*divu + v*var*divv), unith)
182 
.fi
.SS "subroutine physics::fcorlx (real (kind=rk8), dimension(nstu), intent(in) UNITH, integer(kind=ik4), intent(in) NSTU, real (kind=rk8), dimension(:,:), intent(in) F, real (kind=rk8), dimension(:,:), intent(in) V, real (kind=rk8), dimension(:), intent(out) FCX1)"

.PP
Definition at line 188 of file physics\&.f90\&.
.PP
.nf
188 USE portable
189 
190 IMPLICIT NONE
191 INTEGER(KIND=IK4), INTENT(IN)                     :: nstu   ! Number of stations
192 REAL (KIND=RK8), DIMENSION(NSTU), INTENT(IN)      :: unith
193 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: f      ! Coriolis parameter at each station and level\&.
194 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: v      ! V-wind component at each station and level\&.
195 REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)        :: fcx1
196 
197 fcx1    = matmul((v*f/nstu), unith)
198 
.fi
.SS "subroutine physics::fcorly (real (kind=rk8), dimension(nstu), intent(in) UNITH, integer(kind=ik4), intent(in) NSTU, real (kind=rk8), dimension(:,:), intent(in) F, real (kind=rk8), dimension(:,:), intent(in) U, real (kind=rk8), dimension(:), intent(out) FCY1)"

.PP
Definition at line 204 of file physics\&.f90\&.
.PP
.nf
204 USE portable
205 
206 IMPLICIT NONE
207 INTEGER(KIND=IK4), INTENT(IN)                     :: nstu   ! Number of stations
208 REAL (KIND=RK8), DIMENSION(NSTU), INTENT(IN)      :: unith
209 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: f      ! Coriolis parameter at each station and level\&.
210 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: u      ! U-wind component at each station and level\&.
211 REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)        :: fcy1
212 
213 fcy1    = matmul((-u*f/nstu), unith)
214 
.fi
.SS "subroutine physics::fpgd (real (kind=rk8), dimension(:), intent(in) UNITH, real (kind=rk8), dimension(:,:), intent(in) Z, real (kind=rk8), dimension(:,:), intent(in) DZDX, real (kind=rk8), dimension(:), intent(out) FP1)"

.PP
Definition at line 220 of file physics\&.f90\&.
.PP
.nf
220 USE portable
221 USE constants
222 
223 IMPLICIT NONE
224 REAL (KIND=RK8), DIMENSION(:), INTENT(IN)     :: unith
225 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: z
226 REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: dzdx
227 REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)    :: fp1
228 
229 fp1 = matmul((-g*z*dzdx), unith)
230 
.fi
.SS "subroutine physics::height (integer (kind=ik4), intent(in) NP, integer (kind=ik4), intent(in) KS, integer (kind=ik4), intent(in) KT, real (kind=rk8), dimension(np), intent(in) P, real (kind=rk8), dimension(np), intent(in) T, real (kind=rk8), dimension(np), intent(in) R, real (kind=rk8), dimension(np), intent(out) Z)"

.PP
Definition at line 244 of file physics\&.f90\&.
.PP
.nf
244 USE portable
245 USE constants
246 
247 IMPLICIT NONE
248 
249 INTEGER (KIND=IK4), INTENT(IN)                :: np         ! Number of pressure levels in the P, T, R and Z arrays\&.
250 INTEGER (KIND=IK4), INTENT(IN)                :: ks         ! The level which the surface is at\&.
251 INTEGER (KIND=IK4), INTENT(IN)                :: kt         ! Level at which the variational analysis stops\&.
252 REAL (KIND=RK8), DIMENSION(NP), INTENT(IN)    :: p          ! Pressure of each level\&.
253 REAL (KIND=RK8), DIMENSION(NP), INTENT(IN)    :: t          ! Temperature of each level\&.
254 REAL (KIND=RK8), DIMENSION(NP), INTENT(IN)    :: r          ! Mixing ratio at each level?
255 REAL (KIND=RK8), DIMENSION(NP), INTENT(OUT)   :: z          ! Geopotential height\&.
256 
257 !
258 ! Local variables\&.
259 !
260 REAL (KIND=RK8)                               :: ep, r1, dz ! Temporary variables\&.
261 INTEGER (KIND=IK4)                            :: kk         ! Counter\&.
262 
263 ep  = rd/rw
264 
265 DO kk=ks+1,kt
266     r1      = rd*0\&.5*((1+r(kk-1)/ep)/(1+r(kk-1))+(1+r(kk)/ep)/(1+r(kk)))
267     dz      = r1/g*0\&.5*(t(kk-1)/p(kk-1)+t(kk)/p(kk))*(p(kk-1)-p(kk))
268     z(kk)   = z(kk-1)+dz
269 END DO
270 
.fi
.SS "subroutine physics::s_r_to_t_z (real (kind=rk8), dimension(:), intent(in) P, real (kind=rk8), intent(in) PS, real (kind=rk8), intent(in) ZS, real (kind=rk8), dimension(:), intent(in) S, real (kind=rk8), dimension(:), intent(in) R, real (kind=rk8), dimension(:), intent(out) T, real (kind=rk8), dimension(:), intent(out) Z)"

.PP
Definition at line 22 of file physics\&.f90\&.
.PP
.nf
22 USE portable
23 USE constants
24 
25 IMPLICIT NONE
26 REAL (KIND=RK8), DIMENSION(:), INTENT(IN)       :: p    ! Pressures of a vertical column (hPa)
27 REAL (KIND=RK8), INTENT(IN)                     :: ps   ! Surface pressure (hPa)
28 REAL (KIND=RK8), INTENT(IN)                     :: zs   ! Surface geopotential height (m)
29 REAL (KIND=RK8), DIMENSION(:), INTENT(IN)       :: s    ! (K)
30 REAL (KIND=RK8), DIMENSION(:), INTENT(IN)       :: r    ! Mixing ration (g/kg)
31 REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)      :: t    ! Temperature at each level (K)
32 REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)      :: z    ! Geopotential height at each level (m)
33 
34 !
35 ! Local variables\&.
36 !
37 REAL (KIND=RK8)                                 :: g2       ! 2*G
38 REAL (KIND=RK8)                                 :: ep       ! RD/RW
39 INTEGER (KIND=IK4)                              :: ii       ! Counter
40 REAL (KIND=RK8)                                 :: rv, a1, p1, t1, r1, z1, p2, r2
41 
42 !
43 ! Assign and initialise some variables\&.
44 !
45 g2  = 2\&.0*g
46 ep  = rd/rw
47 t   = 0\&.0

src/physics.f90
!***********************************************************************************************************************************
! PHYSICS module
!
! This module contains several subroutines which calculate various physics. One of the useful things about including these
! subroutines in a module is that we don't need to define explicit interfaces for the subroutines.
!
! Tim Hume.
! 2 October 2006.
!***********************************************************************************************************************************
MODULE PHYSICS
CONTAINS

!***********************************************************************************************************************************
! S_R_TO_T_Z
!
! Converted from the IDL procedure of the same name.
!
! Tim Hume.
! 23 August 2006.
!***********************************************************************************************************************************
SUBROUTINE S_R_TO_T_Z(P, PS, ZS, S, R, T, Z)
USE PORTABLE
USE CONSTANTS

IMPLICIT NONE
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)       :: P    ! Pressures of a vertical column (hPa)
REAL (KIND=RK8), INTENT(IN)                     :: PS   ! Surface pressure (hPa)
REAL (KIND=RK8), INTENT(IN)                     :: ZS   ! Surface geopotential height (m)
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)       :: S    ! (K)
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)       :: R    ! Mixing ration (g/kg)
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)      :: T    ! Temperature at each level (K)
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)      :: Z    ! Geopotential height at each level (m)

!
! Local variables.
!
REAL (KIND=RK8)                                 :: G2       ! 2*G
REAL (KIND=RK8)                                 :: EP       ! RD/RW
INTEGER (KIND=IK4)                              :: II       ! Counter
REAL (KIND=RK8)                                 :: RV, A1, P1, T1, R1, Z1, P2, R2

!
! Assign and initialise some variables.
!
G2  = 2.0*G
EP  = RD/RW
T   = 0.0
Z   = 0.0

WHERE (P .GE. PS)
    Z   = ZS
    T   = S(1) - G*ZS/CPD
END WHERE

P1  = PS
T1  = T(1)
R1  = R(1)
Z1  = ZS

DO II=TRANSFER(MAXLOC(P, MASK=P.LT.PS), 0),SIZE(P)
    P2      = P(II)
    R2      = R(II)
    RV      = RD*0.5*((1+R1/EP)/(1+R1)+(1+R2/EP)/(1+R2))
    A1      = RV/2.0/CPD*LOG(P1/P2)
    T(II)   = (S(II) - G/CPD*Z1 - A1*T1)/(1.0 + A1)
    Z(II)   = Z1 + RV*(T1+T(II))/G2*LOG(P1/P2)

    P1      = P2
    R1      = R2
    Z1      = Z(II)
    T1      = T(II)
END DO

END SUBROUTINE S_R_TO_T_Z

!***********************************************************************************************************************************
! T_R_TO_S_Z
!
! This subroutine calculates s and z from p, ps and r.
! It is ported from the IDL procedure with the same name.
!
! Tim Hume.
! 22 September 2006.
!***********************************************************************************************************************************

SUBROUTINE T_R_TO_S_Z(P, PS, ZS, T, R, S, Z)
USE PORTABLE
USE CONSTANTS

IMPLICIT NONE
REAL (KIND=RK8), INTENT(IN)                             :: PS       ! Surface pressure (hPa).
REAL (KIND=RK8), INTENT(IN)                             :: ZS       ! Surface geopotential height (m).
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)               :: P        ! Pressure levels (hPa).
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)               :: T        ! Temperature (K).
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)               :: R        ! Mixing ration (g/kg).
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)              :: S        ! Dry static energy divided by CPD (K).
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)              :: Z        ! Geopotential height (m).

!
! Local variables.
!
REAL (KIND=RK8)                                         :: EP, G2, P1, T1, R1, P2, R2, T2, RV
INTEGER (KIND=IK4)                                      :: LL       ! Counter.

!
! Do some basic checks of the input data. This is an unfortunate necessity (otherwise bad input data can cause overflows in the
! calculations. If bad data are detected, set the geopotential height and dry static energy to an undefined value (-9999.99)
!



IF ((PS .LE. 0) .OR. (MINVAL(P) .LE. 0) .OR. (MINVAL(T) .LT. 0) .OR. (MINVAL(R) .LT. 0)) THEN
    S   = -9999.99
    Z   = -9999.99
    RETURN
END IF

!
! Set some variables which are frequently used.
!
EP  = RD/RW
G2  = 2.0*G

!
! Where P is greater than PS, we are presumably below the surface. Set the geopotential height and dry static energy to surface
! values. We also ensure the geopotential height of the lowest level is the same as the surface.
!

Z(1)    = ZS
Z       = 0
S       = 0

WHERE (P .GE. PS)
    Z   = ZS
    S   = T(1) + G*ZS/CPD
END WHERE

!
! For each non-surface level, calculate the geopotential height and dry static energy.
!
P1  = PS
T1  = T(1)
R1  = R(1)
DO LL=MAX(1,TRANSFER(MAXLOC(P, MASK=P.LT.PS), 0)),SIZE(P)
    P2      = P(LL)
    T2      = T(LL)
    R2      = R(LL)
    RV      = RD*0.5*((1+R1/EP)/(1+R1) + (1+R2/EP)/(1+R2))
    IF (LL.gt.1)  Z(LL)   = Z(LL-1) + RV*(T1+T2)/G2*LOG(P1/P2)
    T1      = T2
    P1      = P2
    R1      = R2
    S(LL)   = T(LL) + G*Z(LL)/CPD
END DO


END SUBROUTINE T_R_TO_S_Z

!*******************************************************************************************************************************
! DIVERG
!
! This subroutine computes the divergence of a field.
! It has been ported from the IDL procedure of the same name.
!
! Tim Hume.
! 22 August 2006.
!*******************************************************************************************************************************

SUBROUTINE DIVERG(UNITH, VAR, U, V, DIVU, DIVV, DIV1)
USE PORTABLE

IMPLICIT NONE
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)     :: UNITH      ! One dimensional array with y columns.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: VAR        ! Array with x columns and y rows containing the variable.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: U          ! Array with x columns and y rows containing U-component of wind.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: V          ! Array with x columns and y rows containing V-component of wind.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: DIVU       ! Array with x columns and y rows containing DIVU.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: DIVV       ! Array with x columns and y rows containing DIVV.
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)    :: DIV1       ! Array with x columns containing divergence.

DIV1    = MATMUL((U*VAR*DIVU + V*VAR*DIVV), UNITH)

END SUBROUTINE DIVERG

!***********************************************************************************************************************************

SUBROUTINE FCORLX(UNITH, NSTU, F, V, FCX1)
USE PORTABLE

IMPLICIT NONE
INTEGER(KIND=IK4), INTENT(IN)                     :: NSTU   ! Number of stations
REAL (KIND=RK8), DIMENSION(NSTU), INTENT(IN)      :: UNITH
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: F      ! Coriolis parameter at each station and level.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: V      ! V-wind component at each station and level.
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)        :: FCX1

FCX1    = MATMUL((V*F/NSTU), UNITH)

END SUBROUTINE FCORLX

!***********************************************************************************************************************************

SUBROUTINE FCORLY(UNITH, NSTU, F, U, FCY1)
USE PORTABLE

IMPLICIT NONE
INTEGER(KIND=IK4), INTENT(IN)                     :: NSTU   ! Number of stations
REAL (KIND=RK8), DIMENSION(NSTU), INTENT(IN)      :: UNITH
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: F      ! Coriolis parameter at each station and level.
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)       :: U      ! U-wind component at each station and level.
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)        :: FCY1

FCY1    = MATMUL((-U*F/NSTU), UNITH)

END SUBROUTINE FCORLY

!***********************************************************************************************************************************

SUBROUTINE FPGD(UNITH, Z, DZDX, FP1)
USE PORTABLE
USE CONSTANTS

IMPLICIT NONE
REAL (KIND=RK8), DIMENSION(:), INTENT(IN)     :: UNITH
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: Z
REAL (KIND=RK8), DIMENSION(:,:), INTENT(IN)   :: DZDX
REAL (KIND=RK8), DIMENSION(:), INTENT(OUT)    :: FP1

FP1 = MATMUL((-G*Z*DZDX), UNITH)

END SUBROUTINE FPGD

!*******************************************************************************************************************************
! HEIGHT
!
! This subroutine computes the geopotential height.
! It has been ported from the IDL procedure of the same name.
!
! Tim Hume.
! 21 August 2006.
!*******************************************************************************************************************************

SUBROUTINE HEIGHT(NP, KS, KT, P, T, R, Z)
USE PORTABLE
USE CONSTANTS

IMPLICIT NONE

INTEGER (KIND=IK4), INTENT(IN)                :: NP         ! Number of pressure levels in the P, T, R and Z arrays.
INTEGER (KIND=IK4), INTENT(IN)                :: KS         ! The level which the surface is at.
INTEGER (KIND=IK4), INTENT(IN)                :: KT         ! Level at which the variational analysis stops.
REAL (KIND=RK8), DIMENSION(NP), INTENT(IN)    :: P          ! Pressure of each level.
REAL (KIND=RK8), DIMENSION(NP), INTENT(IN)    :: T          ! Temperature of each level.
REAL (KIND=RK8), DIMENSION(NP), INTENT(IN)    :: R          ! Mixing ratio at each level?
REAL (KIND=RK8), DIMENSION(NP), INTENT(OUT)   :: Z          ! Geopotential height.

!
! Local variables.
!
REAL (KIND=RK8)                               :: EP, R1, DZ ! Temporary variables.
INTEGER (KIND=IK4)                            :: KK         ! Counter.

EP  = RD/RW

DO KK=KS+1,KT
    R1      = RD*0.5*((1+R(KK-1)/EP)/(1+R(KK-1))+(1+R(KK)/EP)/(1+R(KK)))
    DZ      = R1/G*0.5*(T(KK-1)/P(KK-1)+T(KK)/P(KK))*(P(KK-1)-P(KK))
    Z(KK)   = Z(KK-1)+DZ
END DO

END SUBROUTINE HEIGHT

!*******************************************************************************************************************************
! CALHT2
!
! This subroutine computes ...
! It has been ported from the IDL procedure of the same name.
!
! Tim Hume.
! 14 September 2006.
!*******************************************************************************************************************************

SUBROUTINE CALHT2(PS, ZS, N, P, TC, R, HT, DEW)
USE PORTABLE
USE CONSTANTS

IMPLICIT NONE
REAL (KIND=RK8), INTENT(IN)                       :: PS     ! Surface pressure (units).
REAL (KIND=RK8), INTENT(IN)                       :: ZS     ! Surface geopotential (units).
INTEGER (KIND=IK4), INTENT(IN)                    :: N      ! Number of vertical levels.
REAL (KIND=RK8), DIMENSION(N), INTENT(IN)         :: P      ! Pressures (units).
REAL (KIND=RK8), DIMENSION(N), INTENT(IN)         :: TC     ! Temperatures (Celsius).
REAL (KIND=RK8), DIMENSION(N), INTENT(INOUT)      :: R      ! Relative humidity (%).
REAL (KIND=RK8), DIMENSION(N), INTENT(OUT)        :: HT     ! Geopotential height (units).
REAL (KIND=RK8), DIMENSION(N), INTENT(OUT)        :: DEW    ! Dew-point temperature? (units).

!
! Local variables.
!
REAL (KIND=RK8), DIMENSION(N)                     :: RH     ! Relative humidity converted to 0-1 range.
REAL (KIND=RK8), DIMENSION(N)                     :: T      ! Temperature in Kelvin.
REAL (KIND=RK8), DIMENSION(N)                     :: ES     ! Saturation vapour pressure.
REAL (KIND=RK8), DIMENSION(N)                     :: RS     ! Saturated mixing ratio.
REAL (KIND=RK8)                                   :: EP     ! = RD/RW
INTEGER (KIND=IK4)                                :: KS     ! Index of first vertical level at or above the surface.
REAL (KIND=RK8)                                   :: R1, DZ

!
! Convert the relative humidity to the 0-1 range. Any humidities less than 0 or greater than 1 are set to 1. Not sure of the
! reasoning behind making RH<0 go to 1 ... maybe missing data is better substituted with 1 RH than some other value?
!
RH  = R/100.0               
WHERE ((RH .LT. 0.0) .OR. (RH .GT. 1.0)) RH = 1.0

T   = TC + T0               ! Convert temperatures in Celsius to Kelvin.

!
! Calculate the saturation vapour pressure using a modified form of the Clausius-Clayperon equation. I believe this equation might
! be in Emanuel (1994): "Atmospheric Convection" (maybe).
!
ES  = 1.003*EXP(53.67957 - 6743.769/T - 4.8451*LOG(T))
EP  = RD/RW                 ! This will be used several times in this subroutine (so save resources by only calculating it once).

RS  = EP*ES/(P-ES)          ! Calculate the saturated mixing ratio.

!
! Overwrite the input relative humidities (R, in %) with mixing ratio.
!
R   = RS*RH

!
! Locate the first level, at or above the surface (which has pressure PS). Default to 1 if we can't find the surface.
!
KS  = MAX(TRANSFER(MAXLOC(P, MASK = P .LE. PS), 0),1)

!
! Work out the geopotential height of the first level. I added a little bit of code to deal with the situation where KS=1. The IDL
! code did not seem to deal with this, assuming that there was always at least one pressure level "below the surface", and did a
! simple interpolation. If there is no level below the surface, can't do interpolation.
!
R1      = RD*0.5*((1+R(MAX(1,KS-1))/EP)/(1+R(MAX(1,KS-1))) + (1+R(KS)/EP)/(1+R(KS)))
DZ      = R1/G*0.5*(T(MAX(1,KS-1))/P(MAX(1,KS-1)) + T(KS)/P(KS)) * (PS/P(KS))
HT(KS)  = ZS + DZ

!
! Work out the geopotential height of all the vertical levels above the surface.
!
CALL HEIGHT(N, KS, N, P, T, R, HT)

!
! Assign the surface geopotential height to all levels below the surface.
!
HT(1:MAX(1,KS-1))  = ZS

END SUBROUTINE CALHT2

!***********************************************************************************************************************************
! CALC_STATE2
!
! This subroutine calculates various thermodynamic properties of the atmosphere.
! It is a port of the IDL procedure with the same name.
!
! Tim Hume.
! 23 August 2006.
!***********************************************************************************************************************************

SUBROUTINE CALC_STATE2(P, T, TD, E, Z, R, RH, H, HDRY, S)
USE PORTABLE
USE CONSTANTS

IMPLICIT NONE
REAL (KIND=RK8), INTENT(IN)     :: P    ! Pressure (Pa) - check units.
REAL (KIND=RK8), INTENT(IN)     :: T    ! Temperature (K).
REAL (KIND=RK8), INTENT(IN)     :: TD   ! Dew point temperature (K).
REAL (KIND=RK8), INTENT(IN)     :: E    ! Vapour pressure (Pa - check units.
REAL (KIND=RK8), INTENT(IN)     :: Z    ! Geopotential height (m) - check units.
REAL (KIND=RK8), INTENT(OUT)    :: R    ! Mixing ratio.
REAL (KIND=RK8), INTENT(IN)     :: RH   ! Relative humidity (decimal fraction between 0 and 1). 
REAL (KIND=RK8), INTENT(OUT)    :: H    !
REAL (KIND=RK8), INTENT(OUT)    :: HDRY !
REAL (KIND=RK8), INTENT(OUT)    :: S    !

!
! Local variables.
!
REAL (KIND=RK8)                 :: EP, ES, RS, LV1

EP      = RD/RW
ES      = 1.003*EXP(53.67957 - 6743.769/T - 4.8451*LOG(T))  ! An empirical formula for the saturation vapour pressure.
RS      = EP*ES/(P-ES)
R       = RS*RH
LV1     = LV0 + (CPV-CL)*(T-T0)
H       = (CPD + R*CL)*T + LV1*R + (1.0 + R)*G*Z
HDRY    = CPD*T + G*Z
S       = CPD*LOG(T) - RD*LOG(P) + LV0*R/T - R*RW*LOG(RH)
END SUBROUTINE CALC_STATE2

END MODULE PHYSICS
